Cache resolved variable references

anttijk opened the pull request:
#### 954e3afcfb437134cbc829c3168496bb15f6aa42
<pre>
Cache resolved variable references
<a href="https://bugs.webkit.org/show_bug.cgi?id=261265">https://bugs.webkit.org/show_bug.cgi?id=261265</a>
rdar://115104109

Reviewed by Alan Baradlay.

If the pointed-to variable doesn&apos;t change then the final result will be unchanged.
Also add a fast path for resolving simple var(--foo) reference. Currently the cache
is only used on this path.

* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::CSSVariableReferenceValue):
(WebCore::propertyValueForVariableName):
(WebCore::CSSVariableReferenceValue::resolveVariableReference const):
(WebCore::CSSVariableReferenceValue::cacheSimpleReference):

Detect the simple reference case and cache it.

(WebCore::CSSVariableReferenceValue::tryResolveSimpleReference const):

In the simple case we just grab the underlying value and return it without doing any token manipulation.

(WebCore::CSSVariableReferenceValue::resolveVariableReferences const):

Use the fast path if possible.

(WebCore::CSSVariableReferenceValue::resolveAndCacheValue const):

In case we are depenent of only a single variable data object, check if it is equal to the cache entry
and if so, return the cached final value.

(WebCore::CSSVariableReferenceValue::resolveSingleValue const):
(WebCore::CSSVariableReferenceValue::resolveSubstitutionValue const):

Handle the normal and substitution (shorthand) case using a template.

* Source/WebCore/css/CSSVariableReferenceValue.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::resolveVariableReferences):

Canonical link: <a href="https://commits.webkit.org/267742@main">https://commits.webkit.org/267742@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da74f1e3911985a9cff2a5d1ad1a34aff920d265

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17555 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17883 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18406 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19361 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16424 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17752 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21164 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18028 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18536 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18072 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15246 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20221 "Failed to checkout and rebase branch from PR 17535") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15307 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15983 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22597 "layout-tests (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16316 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16152 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20449 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16727 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14170 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15849 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4179 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20216 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16567 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/css/CSSVariableReferenceValue.cpp:
```diff
@@ -168,13 +172,106 @@ std::optional<Vector<CSSParserToken>> CSSVariableReferenceValue::resolveTokenRan
+            return { };
+
+        return SimpleReference { variableName, functionId };
+    }();
```
`nice { lambda [](){}(); }`

anttijk committed:
Cache resolved variable references
https://bugs.webkit.org/show_bug.cgi?id=261265
rdar://115104109

Reviewed by Alan Baradlay.

If the pointed-to variable doesn't change then the final result will be unchanged.
Also add a fast path for resolving simple var(--foo) reference. Currently the cache
is only used on this path.

* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::CSSVariableReferenceValue):
(WebCore::propertyValueForVariableName):
(WebCore::CSSVariableReferenceValue::resolveVariableReference const):
(WebCore::CSSVariableReferenceValue::cacheSimpleReference):

Detect the simple reference case and cache it.

(WebCore::CSSVariableReferenceValue::tryResolveSimpleReference const):

In the simple case we just grab the underlying value and return it without doing any token manipulation.

(WebCore::CSSVariableReferenceValue::resolveVariableReferences const):

Use the fast path if possible.

(WebCore::CSSVariableReferenceValue::resolveAndCacheValue const):

In case we are depenent of only a single variable data object, check if it is equal to the cache entry
and if so, return the cached final value.

(WebCore::CSSVariableReferenceValue::resolveSingleValue const):
(WebCore::CSSVariableReferenceValue::resolveSubstitutionValue const):

Handle the normal and substitution (shorthand) case using a template.

* Source/WebCore/css/CSSVariableReferenceValue.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::resolveVariableReferences):

Canonical link: https://commits.webkit.org/267742@main
